 
# <font color=blue>COVID-19 Evolution and stats</font>
Victor Salazar 
12061468

The following notebook will try to analyze one the biggest problems that the world is dealing with currently.For this purpose we will focus essentially on which countries and regions are the most affected by the pandemic (COVID-19) since it first appeared, in terms of confirmed cases and number of deaths.

In [1]:
import pandas as pd
import numpy as np 

pd.options.display.max_rows = 10

from math import pi

from bokeh.io import output_notebook, show

from bokeh.transform import transform
from bokeh.palettes import all_palettes

from bokeh.layouts import column, row
from bokeh.plotting import figure
from bokeh.models import CustomJS, Slider, ColumnDataSource, Select, TextInput, BasicTicker, ColorBar,LinearColorMapper, PrintfTickFormatter

from package import date, date_name, daily_increase, monthly, country_to_continent,prueba
from package import heat_df,scatter, line_graph, lines_countries

output_notebook()

np.set_printoptions(precision=4, suppress=True)

Loading BokehJS ...

## Data 

Data has been retreived from "https://www.kaggle.com/imdevskp/corona-virus-report" where data from confirmed cases and deaths from most of the countries in the world has been collected from the Johns Hopkins University database. This data has been constanlty updated (last day included is 27 of April of 2021). 4 data sets can be dowloaded however in this notebook only 2 are considered. One has the cumulative number of confirmed cases and the other has confirmed deaths.   

In [2]:
df = pd.read_csv("time_series_covid_19_confirmed.csv",  parse_dates=True)
df1 =  pd.read_csv("time_series_covid_19_deaths.csv")

From the data set we will only use the cases reported , therefore all the other variables included in the datasets are dropped

In [3]:
df = df.drop(['Province/State','Lat','Long'], axis=1)
df1 =  df1.drop(['Province/State','Lat','Long'], axis=1)

#Countries seem to have multiple entries representing different areas within a same country therefore they are grouped
df = df.groupby('Country/Region').sum()
df1 = df1.groupby("Country/Region").sum()

In [4]:
#columns are transfromed into datetime format to help with the analysis and transformation
df.columns = df.columns.map(date)
df1.columns = df1.columns.map(date)

It can now be appreciated the general format of the dataframes with the total information

In [5]:
display(df)
display(df1)

2020-01-22  2020-01-23  2020-01-24  2020-01-25  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      0           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-26  2020-01-27  2020-01-28  2020-01-29  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      2           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-30  2020-01-31  ...  2021-04-18  2021-04-19  \
Country/Region                              ...                           
Afghanistan                  0           0  ...       57793       57898   
Albania                      0           0  ...      129594      129694   
Algeria                      0           0  ...      119642      119805   
Andorra                      0           0  ...       12805       12805   
Angola                       0           0  ...       24389       24518   
...                        ...         ...  ...         ...         ...   
Vietnam                      2           2  ...        2785        2791   
West Bank and Gaza           0           0  ...      280741      282270   
Yemen                        0           0  ...        5812        5858   
Zambia                       0           0  ...       90918       90942   
Zimbabwe                     0           0  ...       37751       37859   

                    2021-04-20  2021-04-21  2021-04-22  2021-04-23  \
Country/Region                                                       
Afghanistan              58037       58214       58312       58542   
Albania                 129842      129980      130114      130270   
Algeria                 119992      120174      120363      120562   
Andorra                  12874       12917       12942       13007   
Angola                   24661       24883       25051       25279   
...                        ...         ...         ...         ...   
Vietnam                   2800        2812        2824        2830   
West Bank and Gaza      284280      286028      287680      289120   
Yemen                     5918        5960        6020        6056   
Zambia                   91042       91119       91189       91251   
Zimbabwe                 37875       37980       38018       38045   

                    2021-04-24  2021-04-25  2021-04-26  2021-04-27  
Country/Region                                                      
Afghanistan              58730       58843       59015       59225  
Albania                 130409      130537      130606      130736  
Algeria                 120736     

2020-01-22  2020-01-23  2020-01-24  2020-01-25  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      0           0           0           0   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-26  2020-01-27  2020-01-28  2020-01-29  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      0           0           0           0   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-30  2020-01-31  ...  2021-04-18  2021-04-19  \
Country/Region                              ...                           
Afghanistan                  0           0  ...        2539        2546   
Albania                      0           0  ...        2342        2347   
Algeria                      0           0  ...        3155        3160   
Andorra                      0           0  ...         123         123   
Angola                       0           0  ...         561         563   
...                        ...         ...  ...         ...         ...   
Vietnam                      0           0  ...          35          35   
West Bank and Gaza           0           0  ...        3017        3047   
Yemen                        0           0  ...        1126        1132   
Zambia                       0           0  ...        1235        1236   
Zimbabwe                     0           0  ...        1553        1553   

                    2021-04-20  2021-04-21  2021-04-22  2021-04-23  \
Country/Region                                                       
Afghanistan               2549        2557        2561        2565   
Albania                   2353        2358        2364        2367   
Algeria                   3165        3172        3181        3190   
Andorra                    123         123         123         123   
Angola                     565         570         572         574   
...                        ...         ...         ...         ...   
Vietnam                     35          35          35          35   
West Bank and Gaza        3078        3096        3115        3138   
Yemen                     1138        1147        1157        1165   
Zambia                    1236        1238        1240        1245   
Zimbabwe                  1554        1555        1555        1556   

                    2021-04-24  2021-04-25  2021-04-26  2021-04-27  
Country/Region                                                      
Afghanistan               2572        2582        2592        2598  
Albania                   2372        2378        2379        2383  
Algeria                   3198     

In [6]:
# making a copy of the dataframe to have both cumulative and daily increases
df_cumu =df.copy()
df1_cumu =df1.copy()
#Use the function daily_increase to change from cumulative to daily increase
df = daily_increase(df,df_cumu)
df1 = daily_increase(df1,df1_cumu)

In [7]:
#some countries have reported lower cases from one day to another this is corrected by setting a negative value to Nan and then
#filling it with the cases from a day before
df[df<0] = None
df= df.fillna(method='bfill',axis=1)
df1[df1<0] = None
df1= df1.fillna(method='bfill',axis=1)
# After analysing the graphs Turkey has an outlier and India as well , this may be an error in the reporting of the number so it is
# corrected with the actual number found on the internet
df['2020-12-10'].loc['Turkey'] = 30424
df['2021-04-27'].loc['India'] = 362902
df1['2021-04-27'].loc['India']= 3285

After the transformations from cumulative and daily cases reported, the dataframes now have this format

In [8]:
display(df)
display(df_cumu)

2020-01-22  2020-01-23  2020-01-24  2020-01-25  \
Country/Region                                                       
Afghanistan                0.0         0.0         0.0         0.0   
Albania                    0.0         0.0         0.0         0.0   
Algeria                    0.0         0.0         0.0         0.0   
Andorra                    0.0         0.0         0.0         0.0   
Angola                     0.0         0.0         0.0         0.0   
...                        ...         ...         ...         ...   
Vietnam                    0.0         2.0         0.0         0.0   
West Bank and Gaza         0.0         0.0         0.0         0.0   
Yemen                      0.0         0.0         0.0         0.0   
Zambia                     0.0         0.0         0.0         0.0   
Zimbabwe                   0.0         0.0         0.0         0.0   

                    2020-01-26  2020-01-27  2020-01-28  2020-01-29  \
Country/Region                                                       
Afghanistan                0.0         0.0         0.0         0.0   
Albania                    0.0         0.0         0.0         0.0   
Algeria                    0.0         0.0         0.0         0.0   
Andorra                    0.0         0.0         0.0         0.0   
Angola                     0.0         0.0         0.0         0.0   
...                        ...         ...         ...         ...   
Vietnam                    0.0         0.0         0.0         0.0   
West Bank and Gaza         0.0         0.0         0.0         0.0   
Yemen                      0.0         0.0         0.0         0.0   
Zambia                     0.0         0.0         0.0         0.0   
Zimbabwe                   0.0         0.0         0.0         0.0   

                    2020-01-30  2020-01-31  ...  2021-04-18  2021-04-19  \
Country/Region                              ...                           
Afghanistan                0.0         0.0  ...        72.0       105.0   
Albania                    0.0         0.0  ...       138.0       100.0   
Algeria                    0.0         0.0  ...       156.0       163.0   
Andorra                    0.0         0.0  ...        34.0         0.0   
Angola                     0.0         0.0  ...        89.0       129.0   
...                        ...         ...  ...         ...         ...   
Vietnam                    0.0         0.0  ...         4.0         6.0   
West Bank and Gaza         0.0         0.0  ...       988.0      1529.0   
Yemen                      0.0         0.0  ...        42.0        46.0   
Zambia                     0.0         0.0  ...        74.0        24.0   
Zimbabwe                   0.0         0.0  ...        52.0       108.0   

                    2021-04-20  2021-04-21  2021-04-22  2021-04-23  \
Country/Region                                                       
Afghanistan              139.0       177.0        98.0       230.0   
Albania                  148.0       138.0       134.0       156.0   
Algeria                  187.0       182.0       189.0       199.0   
Andorra                   69.0        43.0        25.0        65.0   
Angola                   143.0       222.0       168.0       228.0   
...                        ...         ...         ...         ...   
Vietnam                    9.0        12.0        12.0         6.0   
West Bank and Gaza      2010.0      1748.0      1652.0      1440.0   
Yemen                     60.0        42.0        60.0        36.0   
Zambia                   100.0        77.0        70.0        62.0   
Zimbabwe                  16.0       105.0        38.0        27.0   

                    2021-04-24  2021-04-25  2021-04-26  2021-04-27  
Country/Region                                                      
Afghanistan              188.0       113.0       172.0       210.0  
Albania                  139.0       128.0        69.0       130.0  
Algeria                  174.0     

2020-01-22  2020-01-23  2020-01-24  2020-01-25  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      0           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-26  2020-01-27  2020-01-28  2020-01-29  \
Country/Region                                                       
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      2           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-30  2020-01-31  ...  2021-04-18  2021-04-19  \
Country/Region                              ...                           
Afghanistan                  0           0  ...       57793       57898   
Albania                      0           0  ...      129594      129694   
Algeria                      0           0  ...      119642      119805   
Andorra                      0           0  ...       12805       12805   
Angola                       0           0  ...       24389       24518   
...                        ...         ...  ...         ...         ...   
Vietnam                      2           2  ...        2785        2791   
West Bank and Gaza           0           0  ...      280741      282270   
Yemen                        0           0  ...        5812        5858   
Zambia                       0           0  ...       90918       90942   
Zimbabwe                     0           0  ...       37751       37859   

                    2021-04-20  2021-04-21  2021-04-22  2021-04-23  \
Country/Region                                                       
Afghanistan              58037       58214       58312       58542   
Albania                 129842      129980      130114      130270   
Algeria                 119992      120174      120363      120562   
Andorra                  12874       12917       12942       13007   
Angola                   24661       24883       25051       25279   
...                        ...         ...         ...         ...   
Vietnam                   2800        2812        2824        2830   
West Bank and Gaza      284280      286028      287680      289120   
Yemen                     5918        5960        6020        6056   
Zambia                   91042       91119       91189       91251   
Zimbabwe                 37875       37980       38018       38045   

                    2021-04-24  2021-04-25  2021-04-26  2021-04-27  
Country/Region                                                      
Afghanistan              58730       58843       59015       59225  
Albania                 130409      130537      130606      130736  
Algeria                 120736     

In order to know which countries have had the most cases confirmed so far we need to use the cumulative dataset and sort the values , resuting in the following

In [9]:
df_cumu.sort_values(by=['2021-04-27'], ascending=False).head(10)

2020-01-22  2020-01-23  2020-01-24  2020-01-25  2020-01-26  \
Country/Region                                                               
US                       1           1           2           2           5   
India                    0           0           0           0           0   
Brazil                   0           0           0           0           0   
France                   0           0           2           3           3   
Russia                   0           0           0           0           0   
Turkey                   0           0           0           0           0   
United Kingdom           0           0           0           0           0   
Italy                    0           0           0           0           0   
Spain                    0           0           0           0           0   
Germany                  0           0           0           0           0   

                2020-01-27  2020-01-28  2020-01-29  2020-01-30  2020-01-31  \
Country/Region                                                               
US                       5           5           6           6           8   
India                    0           0           0           1           1   
Brazil                   0           0           0           0           0   
France                   3           4           5           5           5   
Russia                   0           0           0           0           2   
Turkey                   0           0           0           0           0   
United Kingdom           0           0           0           0           2   
Italy                    0           0           0           0           2   
Spain                    0           0           0           0           0   
Germany                  1           4           4           4           5   

                ...  2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Country/Region  ...                                                   
US              ...    31670031    31737964    31799237    31862094   
India           ...    15061805    15320972    15616130    15930774   
Brazil          ...    13943071    13973695    14043076    14122795   
France          ...     5350521     5357229     5401305     5436229   
Russia          ...     4649044     4657509     4665553     4673699   
Turkey          ...     4268447     4323596     4384624     4446591   
United Kingdom  ...     4403060     4406114     4408644     4411068   
Italy           ...     3870131     3878994     3891063     3904899   
Spain           ...     3407283     3428354     3428354     3446072   
Germany         ...     3155522     3167137     3198534     3222888   

                2021-04-22  2021-04-23  2021-04-24  2021-04-25  2021-04-26  \
Country/Region                                                               
US                31929351    31991750    32045113    32077178    32124869   
India             16263695    16610481    16960172    17313163    17636186   
Brazil            14167973    14237078    14308215    14340787    14369423   
France             5469674     5502014     5534656     5559121     5565080   
Russia             4682573     4691290     4699988     4708640     4717321   
Turkey             4501382     4550820     4591416     4629969     4667281   
United Kingdom     4413834     4416588     4418710     4420443     4422562   
Italy              3920945     3935703     3949517     3962674     3971114   
Spain              3456886     3468617     3468617     3468617     3488469   
Germany            3254609     3276871     3291293     3306692     3312653   

                2021-04-27  
Country/Region              
US                32175049  
India             17636307  
Brazil            14441563  
France             5595403  
Russia             4725252  
Turkey             4710582  
United Kingdom     4425259  
Italy              3981512  
Spain              3496134  
Germany            3338564  

[10

By total number of cases this are the top 10 countries with the highest total cases since the pandemic started

In [10]:
df1_cumu.sort_values(by=['2021-04-27'], ascending=False).head(10)

2020-01-22  2020-01-23  2020-01-24  2020-01-25  2020-01-26  \
Country/Region                                                               
US                       0           0           0           0           0   
Brazil                   0           0           0           0           0   
Mexico                   0           0           0           0           0   
India                    0           0           0           0           0   
United Kingdom           0           0           0           0           0   
Italy                    0           0           0           0           0   
Russia                   0           0           0           0           0   
France                   0           0           0           0           0   
Germany                  0           0           0           0           0   
Spain                    0           0           0           0           0   

                2020-01-27  2020-01-28  2020-01-29  2020-01-30  2020-01-31  \
Country/Region                                                               
US                       0           0           0           0           0   
Brazil                   0           0           0           0           0   
Mexico                   0           0           0           0           0   
India                    0           0           0           0           0   
United Kingdom           0           0           0           0           0   
Italy                    0           0           0           0           0   
Russia                   0           0           0           0           0   
France                   0           0           0           0           0   
Germany                  0           0           0           0           0   
Spain                    0           0           0           0           0   

                ...  2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Country/Region  ...                                                   
US              ...      567344      567825      568649      569491   
Brazil          ...      373335      374682      378003      381475   
Mexico          ...      212339      212466      213048      213597   
India           ...      178769      180530      182553      184657   
United Kingdom  ...      127518      127524      127557      127577   
Italy           ...      116927      117243      117633      117997   
Russia          ...      103834      104173      104545      104937   
France          ...      100892      101339      101713      102046   
Germany         ...       80052       80353       80680       80938   
Spain           ...       76981       77102       77102       77364   

                2021-04-22  2021-04-23  2021-04-24  2021-04-25  2021-04-26  \
Country/Region                                                               
US                  570434      571197      571921      572200      572674   
Brazil              383502      386416      389492      390797      391936   
Mexico              214095      214504      214853      214947      215113   
India               186920      189544      192311      195123      197894   
United Kingdom      127597      127638      127670      127681      127688   
Italy               118357      118699      119021      119238      119539   
Russia              105328      105718      106108      106434      106783   
France              102323      102655      102872      103017      103415   
Germany              81203       81492       81610       81671       82009   
Spain                77496       77591       77591       77591       77738   

                2021-04-27  
Country/Region              
US                  573378  
Brazil              395022  
Mexico              215547  
India               197894  
United Kingdom      127705  
Italy               119912  
Russia              107167  
France              103762  
Germany              82325  
Spain                77855  

[10

Slight differences like the inclusion of Mexico in the top 10 highest death reported list is appreciated

From all the countries is useful to make a dictionary in order to separate them by continent or groups as top 10 or biggest population (values taken from the internet)

In [11]:
from collections import defaultdict
#Creating a default dictionary in order to have different continets with lists of countries
continents = defaultdict(list)
# Some countries have different names that are not recognized by the module pycountry_convert therefore they needd to be added manually
diff= []

countries= list(monthly(df_cumu).columns)
for x in countries:
    continents["All"].append(x)
    try:
        if country_to_continent(x) == 'Asia':
            continents[country_to_continent(x)].append(x)
        if country_to_continent(x) == 'Europe':
            continents[country_to_continent(x)].append(x)
        if country_to_continent(x) == 'North America':
            continents[country_to_continent(x)].append(x)
        if country_to_continent(x) == 'South America':
            continents[country_to_continent(x)].append(x)
        if country_to_continent(x) == 'Oceania':
            continents[country_to_continent(x)].append(x)
        if country_to_continent(x) == 'Africa':
            continents[country_to_continent(x)].append(x)
    
    # skip the error we already know is beacuse the names have different wording
    except KeyError :
        diff.append(x)
    except TypeError :
        diff.append(x)
    
# manually adding the different country names 
continents['Asia'].append(diff[0])
continents['Asia'].append(diff[9])
continents['Africa'].append(diff[1])
continents['Africa'].append(diff[2])
continents['Africa'].append(diff[3])
continents['North America'].append(diff[11])
continents['Most populated'] = ['China','India','US','Indonesia','Pakistan','Brazil','Nigeria','Bangladesh','Russia','Mexico']
continents['Top 10'] = ['US','India','Brazil','Russia','United Kingdom','France','Spain','Italy','Turkey','Germany']

## Visualization

We will use the cumulative and the daily cases datasets to visuallize in a heatmap, line map and a scatter plot which countries has been the most affected, however many countries have drastic differences in population (consider China and The Netherlands), this is the reason of why countries are grouped by continents and by biggest population and most cases.

### Confirmed cases


In order to analyze in a better way how COVID-19 evolved confirmed cases are gonna be first and then Deaths.
Cumulative and Daily increase show a different picture of the reality.

In [12]:
heat_df(df_cumu,continents) 

It is visible that USA, India and Brazil has been the countries which has been the most affected in the world having the highest cases. US has been the country with most reported cases in total since the pandemic started. In Europe France, Germany,Italy, Spain and Russia are amongst the most affected. In Africa South Africa is the most affected by far. An intresting fact is that China even while having a bigger population than the US and China report way less cases being less affected than other countries of the region and of the world.


In [13]:
heat_df(df,continents) 

If we display a daily reported cases dataframe the picture changes. India is the current country witht he highest cases reported in the world by day. The US has seen a considerable reduction of the cases reported , while Brazil has a steady number reported since July 2020, with a small increase during the last months. For the US the worse months have been December 2020 and January 2021. Australia is the most affected in Oceania but has seen a considerable decrease in their number of cases reported. A curious increase is the one in Papua New Guinea.

In [14]:
line_graph(df)

It is intresting to analysie every country alone. It is possible to appreciate the huge increase in cases in India, or the current decrease in cases in the US. Check your country to see the evolution 

In [15]:
lines_countries(df,continents,'Top 10','Most populated')

It is extremely useful to compare in a same graph the most affected countries, it can be appreciated how India is reporting the highest number of cases replacing the US as the country with most cases reported

### Confirmed deaths


Confirmed cases do affect a country, pushing govrnments to declare lockdowns, curfews or restrictions of movility. However cases itself are not the most damaging to a country, the lifes lost are a bigger problem. That is why is also important to analyse in this specific sense 

In [16]:
heat_df(df1_cumu,continents) 

A similar picture is showed with US , Brazil and India beign the most affected. Nevertheless some countries such as Mexico now appear within the highest deaths in the world. Some Europeans have reported high number of fatalities. France, Italy, Spain, Russia within the most affcted.

In [17]:
heat_df(df1,continents) 

If we look at the daily report of deaths , Brazil is currently the most affected, with India showing a high increase in their fatalities. 

In [18]:
line_graph(df1)

It can be appreciated the drasctic increase in India, and the highest cases reported constantly by Brazil 

In [19]:
lines_countries(df1,continents,'Top 10','Most populated')

In the case of fatalities reported Brazil and the US have reported the most cases and the difference with other countries is quite big. India is showing high deaths figures in the last days

It is intresting to observe the evolution of the cases from countries, pick a region and play with the number of days since the pandemic started and check the evolution

In [20]:
scatter(df_cumu,df1_cumu,continents)

In [21]:
print(df_cumu['2021-04-27'].sum())
print(df1_cumu['2021-04-27'].sum())

148355390
3131660


Finally since the COVID-19 pandemic started, there has been 148,355,390 cases reported till April 27 of May and 3,131,660 reported fatalities